witness.md

```diff
@@ -0,0 +1,599 @@
+
+```
+
+### `NEW_TRIE`
+
+No substitution rules. This instruction is used as a "divider" when building
+a forest of tries.
+
+### `BRANCH mask`
+
+Replaces `NBITSET(mask)` `Node`s to the left of the instruction with a single
+`BranchNode` with these nodes as children according to `mask`.
+
+**Substitution rules**
+```
+
+GUARD NBITSET(mask) == 2
+
+Node(n0) Node(n1) BRANCH(mask) |=> 
+BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
+
+---
+
+GUARD NBITSET(mask) == 3
+
+Node(n0) Node(n1) Node(n2) BRANCH(mask) |=> 
+BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1, n2)}
+
+---
+
+...
+
+---
+
+GUARD NBITSET(mask) == 16
+
+Node(n0) Node(n1) ... Node(n15) BRANCH(mask) |=>
+BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1, ..., n15)}
+```
+
+## Helper functions
+
+### `MAKE_VALUES_ARRAY`
+
+returns an array of 16 elements, where values from `values` are set to the indices where `mask` has bits set to 1. Every other place has `nil` value there.
+
+**Example**: `MAKE_VALUES_ARRAY(5, [a, b])` returns `[a, nil, b, nil, nil, ..., nil]` (binary representation of 5 is `0000000000000101`)
+
+```
+MAKE_VALUES_ARRAY(mask, values...) {
+    return MAKE_VALUES_ARRAY(mask, 0, values)
+}
+
+MAKE_VALUES_ARRAY(mask, idx, values...) {
+    if idx > 16 {
+        return []
+    }
+
+    if BIT_TEST(mask, idx) {
+        return PREPEND(FIRST(values), (MAKE_VALUES_ARRAY mask, INC(idx), REST(values)))
+    } else {
+        return PREPEND(nil, (MAKE_VALUES_ARRAY mask, INC(idx), values))
+    }
+}
+```
+
+
+### `RLP(value)`
+
+returns the RLP encoding of a value
+
+
+### `NBITSET(number)`
+
+returns number of bits set in the binary representation of `number`.
+
+### `BIT_TEST(number, n)`
+
+`n` MUST NOT be negative.
+
+returns `true` if bit `n` in `number` is set, `false` otherwise.
+
+### `PREPEND(value, array)`
+
+returns a new array with the `value` at index 0 and `array` values starting from index 1
+
+### `INC(value)`
+
+increments `value` by 1
+
+### `FIRST(array)`
+
+returns the first value in the specified array
+
+### `REST(array)`
+
+returns the array w/o the first item
+
+### `KECCAK(bytes)`
+
+returns a keccak-256 hash of `bytes`
+
+
+## Serialization
+
+The format for serialization of everything except hashes (that we know the
+length of) is [CBOR](https://cbor.io). It is RFC-specified and concise.
+
+For hashes we use the optimization of knowing the lengths, so we just read 32
+bytes
+
+### Block Witness Format
+
+Each block witness consists of a header followed by a list of instructions.
+
+There is no length of witness specified anywhere, the code expects to just reach `EOF`.
+
+Serialized Witness: `(HEADER, OP1, OP2, ..., OPn-1, OPn, EOF)`
+
+#### Encoding
+
+##### CBOR
+
+The parts of the key that are encoded with CBOR are marked by the `CBOR` function.
+
+##### Keys
+
+Keys are also using custom encryption to make them more compact.
+
+The nibbles of a key are encoded in a following way `[FLAGS NIBBLE1+NIBBLE2 NIBBLE3+NIBBLE4 NIBBLE5... ]`
```

Comment:
I am guessing that square brackets means concatenation. The Serialized Witness above using parentheses is also concatenation of bytes.

Reply:
yeah, that is a typo from the previous version, thanks for noticing it!